semgrep/semgrep/join_rule.py

```diff
@@ -0,0 +1,404 @@
+import json
+import logging
+import sys
+import tempfile
+from collections import defaultdict
+from enum import Enum
+from functools import reduce
+from itertools import chain
+from pathlib import Path
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+from typing import Type
+
+import attr
+import peewee as pw
+from peewee import ModelSelect
+from ruamel.yaml import YAML
+
+import semgrep.semgrep_main
+from semgrep.config_resolver import Config
+from semgrep.config_resolver import resolve_config
+from semgrep.error import ERROR_MAP
+from semgrep.error import FATAL_EXIT_CODE
+from semgrep.error import Level
+from semgrep.error import SemgrepError
+from semgrep.pattern_match import PatternMatch
+from semgrep.rule import Rule
+from semgrep.rule_match import RuleMatch
+
+yaml = YAML()
+
+logger = logging.getLogger(__file__)
+logger.setLevel(logging.INFO)
+handler = logging.StreamHandler(stream=sys.stderr)
+handler.setFormatter(
+    logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+)
+logger.addHandler(handler)
+
+# TODO: refactor into nice code files instead of this giant file
+# TODO: probably, add error handling
+# TODO: decide how to represent these kinds of rules in the output.
+# TODO(bug): join rules don't propagate metavariables forward into messages
+# # report the last finding? report multiple findings?
+
+
+class InvalidConditionError(SemgrepError):
+    level = Level.ERROR
+    code = FATAL_EXIT_CODE
+
+
+### Utility functions
+def group(items: List[Any], key: Callable[[Any], Any]) -> Dict[Any, Any]:
+    dd = defaultdict(list)
+    for item in items:
+        k = key(item)
+        dd[k].append(item)
+    return dd
+
+
+def camel_case(s: str) -> str:
+    return "".join(c for c in s.title() if c.isalnum())
+
+
+class Operator(Enum):
```

Comment:
We have another Operator class elseswhere so probably better for sanity to just have this be JoinOperator